proposals/0000-Move-Platforms-to-Own-Packages.md

```diff
@@ -0,0 +1,273 @@
+   * `ViewPagerAndroid`
+   * **APIs**
+   * `DatePickerAndroid`
+   * `PermissionsAndroid`
+   * `TimePickerAndroid`
+
+
+### Hypothetical project overview (non-exhaustive)
+
+This is the *very basic* overview of what the GitHub repo could look like:
+
+```
+-| project root
+--📁 {.circleci, .github, bots}/
+--📁 flow/
+--📁 flow-github/
+--📁 packages/
+---📁 react-native/
+----📁 Libraries/ (All cross-platform JS)
+----📁 ReactCommon/
+----📁 scripts/
+-----📄 launchPackager.bat
+-----📄 launchPackager.command
+-----📄 packager.sh
+----📁 third-party/
+----📁 tools/ (contains bzl build-defs)
+---📁 react-native-ios/
+----📁 Libraries/
+----📁 React/
+----📁 scripts/
+-----📄 ios-configure-glog.sh
+-----📄 ios-install-third-party.sh
+-----📄 react-native-xcode.sh
+----📁 third-party-podspecs/
+---📁 react-native-android/
+----📁 gradle/
+----📁 keystores/
+----📁 Libraries/
+----📁 ReactAndroid/
+---📁 react-native-git-upgrade/
+--📁 RNTester/
+--📁 scripts/ (CI, sync, etc. scripts)
+--📄 package.json (for the Yarn workspace)
+--📄 {readme, contributing, CoC, other markdown docs}
+```
+
+I am not fully sure what the layout of the local-cli stuff will end up needing to be after this; with [this proposal](https://github.com/react-native-community/discussions-and-proposals/blob/master/proposals/0002-Extracted-Command-Line-Tools.md) the bulk of it will end up getting moved to another repo anyways.
+
+ReactIOS and ReactAndroid would get entries in their `package.json` like this:
+
+```json
+{
+  "rnpm": {
+    "haste": {
+      "providesModuleNodeModules": ["react-native-ios"],
+      "platforms": ["ios"]
+    }
+  }
+}
+```
+
+Platform-specific Flow prop types will probably need to stay bundled together in the main package. Ideally it could be set up so if you didn't have `react-native-ios` installed the iOS-only View props wouldn't show up when typechecking, for example. But I don't know if there's a way to do that with Flow yet. Would need [this issue](https://github.com/facebook/flow/issues/396) to be resolved, probably.
+
+The native build scripts (gradle, BUCK, xcode, etc.) will all need to be updated for the new code locations. Great care must be taken to not break the existing large ecosystem of native compnents on NPM!
+
+## Drawbacks
+
+ - A lot of old open pull requests will need to be rebased if the codebase is re-arranged
+ - The `react-native-ios` and `react-native-android` packages on NPM already have the following versions on them:
+   > - `react-native-ios`: `1.0.0`, `1.0.1`, `1.1.0`
+   > - `react-native-android`: `1.0.0`, `2.0.0`, `3.0.0`
+   
+   This will make React Native's future semver bump to `1.0.0` awkward, if not impossible. Unless it does what React did and jumps to version `60.0.0` or something like that. Or starts at version `4.0.0`, which seems kinda random. NPM also says that for various technical and security reasons, these versions can not be deleted and re-used. I've `npm deprecate`d them, which is the best I can do sadly.
```

Comment:
That's definitely an option! That idiom would make it awkward for `react-native-dom` however, since `react-dom` already exists.

Reply:
I think `react-dom` as a package name makes sense for what it does, and wouldn't make sense as a name for a package in this context. We'd only want a single web implementation, so I don't think we need to worry about mapping every OSS RN variant to an equivalent package in RN.

Reply:
Def. worth making another issue for, since your idea of having information about out-of-tree platforms in the main repo could probably be done independently of this proposal.

> We'd only want a single web implementation

RNWeb and RNDom are different enough projects I think there's room for the both of them. But I really don't know about dropping the `native` from the name, because a "React renderer" and a "React Native renderer/platform" are very different things.

Reply:
what about going babel’s route and going namespaced.

@react-native/core
@react-native/ios
@react-native/android
@react-native/windows
@react-native/tvos

it would allow for semver 1.0.0 and identify official packages ie react-native-web would immediately be identifiable as an external project.

Reply:
@empyrical I ran a namespace search and it appears the orgid [react-native](https://www.npmjs.com/~react-native) is taken (0 public releases may be private only) but orgs [react](https://www.npmjs.com/~react) and [reactnative](https://www.npmjs.com/~reactnative) are not. Maybe worth snagging them now even if they will not be used.

Reply:
no worries! FWIW, facebook DOES own `@react` now. I could see if the owner of `@reactnative` would be open to transferring it, but minus the dash `-` is a bit weird

Reply:
I couldn't see who owned @react because the profile was empty. I reached out to npm asking to be pointed to who has both @react and @reactnative earlier, pointing them here.

If @react is now under facebooks control, if they added both the react and react native community to the org... this would be a great naming convention if everyone got on board:

@react/core
@react/native-core
@react/prop-types
@react/dom
@react/ios
@react/android
@react/web
@react/ etc etc etc

Very straightforward.

Reply:
I've learned that RN and React might be converging a bit more, so that (and @necolas' suggestion to use `react-ios` as well) would make sense 👍 

I've updated the proposal